### PR TITLE
add real demotion and promotion option

### PR DESCRIPTION
**Pull request by VladimirWork**
Signed-off-by: Vladimir Shishkin <vladimir.shishkin@dsr-corporation.com>

### CODE REVIEW

File: scripts/performance/perf_load/perf_processes.py [hunk 1]
```diff
@@ -448,6 +453,73 @@ def check_genesis(gen_path):
...
+            _data['services'] = ['VALIDATOR']
+            req = await ledger.build_node_request(req_data['steward_did'], req_data['node_dest'], json.dumps(_data))
+            await ledger.sign_and_submit_request(pool_handle, wallet_handle, req_data['steward_did'], req)
+            # restart promoted node
+            host = testinfra.get_host('ssh://persistent_node' + req_data['node_alias'][4:])
+            host.run('sudo systemctl restart indy-node')
+
+    # https://stackoverflow.com/questions/37512182/how-can-i-periodically-execute-a-function-with-asyncio
+    task = loop.create_task(_demote_promote_periodic())
+    loop.call_later(interval * 1000, task.cancel)  # define delay before looping task stopping
```
Comment (ashcherbakov) [inline scripts/performance/perf_load/perf_processes.py]: It cancels the demotion-promotion loop in `interval * 1000`, doesn't it? Is it intended? Why don't run promotion-demotion until the test is stopped?

File: scripts/performance/perf_load/perf_processes.py [hunk 2]
```diff
@@ -448,6 +453,73 @@ def check_genesis(gen_path):
...
+
+    try:
+        loop.run_until_complete(task)
```
Comment (ashcherbakov) [inline scripts/performance/perf_load/perf_processes.py]: It waits until the `_demote_promote_periodic` is completed, doesn't it? So, no other requests will be sent in between?

File: scripts/performance/perf_load/perf_processes.py [hunk 3]
```diff
@@ -448,6 +453,73 @@ def check_genesis(gen_path):
...
+            await asyncio.sleep(interval)
+
+            _data['services'] = ['VALIDATOR']
+            req = await ledger.build_node_request(req_data['steward_did'], req_data['node_dest'], json.dumps(_data))
+            await ledger.sign_and_submit_request(pool_handle, wallet_handle, req_data['steward_did'], req)
+            # restart promoted node
+            host = testinfra.get_host('ssh://persistent_node' + req_data['node_alias'][4:])
+            host.run('sudo systemctl restart indy-node')
+
+    # https://stackoverflow.com/questions/37512182/how-can-i-periodically-execute-a-function-with-asyncio
```
Comment (ashcherbakov) [inline scripts/performance/perf_load/perf_processes.py]: It's not good to mention stackoverflow in code/comments

File: scripts/performance/perf_load/perf_processes.py [hunk 4]
```diff
@@ -448,6 +453,73 @@ def check_genesis(gen_path):
         shutil.rmtree(dir_to_dlt, ignore_errors=True)
 
 
+def demote_promote(gen_path, interval):
```
Comment (ashcherbakov) [inline scripts/performance/perf_load/perf_processes.py]: Why don't we send it like a common request sending in load tests?

Comment (VladimirWork) [reply scripts/performance/perf_load/perf_processes.py]: Reworked according to discussion.

**Comment by VladimirWork** [comment]
test this please

**Comment by VladimirWork** [comment]
test this please

File: scripts/performance/perf_load/perf_utils.py [hunk 5]
```diff
@@ -130,3 +131,51 @@ def gen_input_output(addr_txos, val):
...
+
+        # put all into list of dicts
+        self._pool_data = [
+            {'node_alias': node_alias,
+             'node_dest': node_dest}
+            for node_alias, node_dest in zip(aliases, dests)
+        ]
+
+    # pick random node from pool and set it as current (for demotion)
+    def get_node(self):
```
Comment (ashcherbakov) [inline scripts/performance/perf_load/perf_utils.py]: I would call it `select_new_random_node`

File: scripts/performance/perf_load/perf_processes.py [hunk 6]
```diff
@@ -101,12 +102,15 @@
 parser.add_argument('--taa_version', default="test_taa", type=str, required=False,
                     help='Transaction author agreement version')
 
+parser.add_argument('--promotion_shift', default=0, type=int, required=False,
```
Comment (ashcherbakov) [inline scripts/performance/perf_load/perf_processes.py]: Should it be `0` be default? I think something like 60 sec is a better default value.

File: scripts/performance/perf_load/perf_gen_req_parser.py [hunk 7]
```diff
@@ -23,7 +25,7 @@ class ReqTypeParser:
          "get_revoc_reg_delta": RGGetRevocRegDelta, "get_payment_sources": RGGetPaymentSources, "payment": RGPayment,
          "verify_payment": RGVerifyPayment, "cfg_writes": RGConfigChangeState, "demoted_node": RGPoolNewDemotedNode,
          "get_txn": RGGetTxn, 'get_taa': RGGetTAA, 'get_taa_aml': RGGetTAAAML, 'get_auth_rule': RGGetAuthRules,
-         "get_fees": RGGetFees}
+         "get_fees": RGGetFees, "demote": RGDemoteNode, "promote": RGPromoteNode}
```
Comment (ashcherbakov) [inline scripts/performance/perf_load/perf_gen_req_parser.py]: Please update Readme with the info about new requests. Please mention there how to use them (that rate needs to equal, and promotion shift must be set)

File: scripts/performance/perf_load/perf_utils.py (hunk 5 above)
Comment (VladimirWork) [reply scripts/performance/perf_load/perf_utils.py]: Done.

File: scripts/performance/perf_load/perf_processes.py (hunk 6 above)
Comment (VladimirWork) [reply scripts/performance/perf_load/perf_processes.py]: Done.

File: scripts/performance/perf_load/perf_gen_req_parser.py (hunk 7 above)
Comment (VladimirWork) [reply scripts/performance/perf_load/perf_gen_req_parser.py]: Done.

**Pull request merged**